### PR TITLE
[X86] Use RORX over SHR imm

### PR DESCRIPTION
**Pull request by Bryce-MW**
SHRX is preferred over SHR to avoid setting flags but only for variable shifts. If the output of SHR is being truncated, and the immediate shift is less than the number of bits in the source register minus the number of bits in the result, RORX can be used instead. The most common case would be extracting the top half of a register. I could also see it being used when extracting a byte from a larger register.

I am new to tablegen so I am sure this is not being done in the best way.

As far as I can tell, rorx has the same performance characteristics as shr other than not impacting flags.

The following example was my motivation for doing this:
```c
#include <immintrin.h>

unsigned short checksum(const int* data) {
  const int len = 5;
  unsigned out = data[0];
  unsigned int carry = 0;

  #pragma clang loop unroll(enable)
  for (unsigned int i = 1; i < len; i++) {
    out = __builtin_addc(out, data[i], carry, &carry);
  }
  out = __builtin_addcs((unsigned short)out, (unsigned short)(out >> 16), (unsigned short)carry, (unsigned short*)&carry);
  out += carry;
  return ~(unsigned short)out;
}
```
Currently produces:
```asm
checksum:
        mov     ecx, dword ptr [rdi]
        add     ecx, dword ptr [rdi + 4]
        adc     ecx, dword ptr [rdi + 8]
        adc     ecx, dword ptr [rdi + 12]
        adc     ecx, dword ptr [rdi + 16]
        setb    dl
        mov     eax, ecx
        shr     eax, 16
        add     dl, 255
        adc     ax, cx
        adc     ax, 0
        not     eax
        ret
```
With these changes, it produces:
```asm
checksum:
        mov     ecx, dword ptr [rdi]
        add     ecx, dword ptr [rdi + 4]
        adc     ecx, dword ptr [rdi + 8]
        adc     ecx, dword ptr [rdi + 12]
        adc     ecx, dword ptr [rdi + 16]
        rorx    eax, ecx, 16
        adc     ax, cx
        adc     ax, 0
        not     eax
        ret
```

EDIT: Messed up git as usual. I've fixed the broken tests. Need to add new ones. I don't feel super comfortable with the SHR by 1 to RORX changes. Might want to exclude that? Or maybe not because SHR by 1 still sets flags. Maybe some detection should be done on if making transformations like this are actually worthwhile

### CODE REVIEW

File: llvm/lib/Target/X86/X86InstrShiftRotate.td [hunk 1]
```diff
@@ -879,6 +879,26 @@ let Predicates = [HasBMI2, HasEGPR, In64BitMode] in {
   defm SHLX64 : bmi_shift<"shlx{q}", GR64, i64mem, "_EVEX">, T8, PD, REX_W, EVEX;
 }
 
+
+def immle16_8 : ImmLeaf<i8, [{
```
Comment (topperc) [inline llvm/lib/Target/X86/X86InstrShiftRotate.td:883]: maybe `immle16minus8` to be more clear?

Comment (Bryce-MW) [reply llvm/lib/Target/X86/X86InstrShiftRotate.td:883]: Good idea. I was trying to make something that took two parameters (e.g. `immle<16, 8>`) but I couldn't quite figure that out

File: llvm/test/CodeGen/X86/cmp-shiftX-maskX.ll [hunk 2]
```diff
@@ -157,7 +157,8 @@ define i1 @shr_to_shl_eq_i16_s1_fail(i16 %x) {
 ; CHECK:       # %bb.0:
 ; CHECK-NEXT:    movzwl %di, %eax
 ; CHECK-NEXT:    andl $32766, %edi # imm = 0x7FFE
-; CHECK-NEXT:    shrl %eax
+; CHECK-NOBMI-NEXT:    shrl %eax
```
Comment (phoebewang) [inline llvm/test/CodeGen/X86/cmp-shiftX-maskX.ll:160]: Use `utils/update_llc_test_checks.py` to update the test.

Comment (Bryce-MW) [reply llvm/test/CodeGen/X86/cmp-shiftX-maskX.ll:160]: I knew I was missing something. Thank you!

**Review by phoebewang — APPROVED** [review]
LGTM with two nits.

**Review by KanRobert — CHANGES_REQUESTED** [review]
Sorry for request change on this. I have two concerns here
1.  Whether this change will bring benefits is questionable b/c
      a.  it introduces false dependency 
      b.  code size of RORX is longer than SHR
2.  it's probably not the best solution to do this by adding new pattern. This 
      optimization should occur at most when there is a user in the previously generated 
      EFLAGS.

**Comment by Bryce-MW** [comment]
I appreciate the comment! I actually was thinking about some of your points but I wasn't sure how much of an issue they were so I am glad that you brought them up.

I think this is definitely a benefit for the flag spilling case. Probably worthwhile even when flags other than those set by SHR are read.

But the code size is a concern. Especially when replacing SHR by 1 which is 2 bytes vs 6 for RORX. Even the MOV + full SHR is 5 bytes and I don't know that there is a benefit to eliminating a MOV in most cases. Obviously depending on REX and whatever but I think it's generally worse?

Changing and 8bit or 16bit shift to a 32bit RORX could have a false dependency so I'd definitely remove those. Thanks for pointing that out. Otherwise it should be fine?

I have two questions, first is where this should better be done. I haven't looked at codegen much before so I'm not so sure on where things happen. I also think there may be other situations where a normally less optimal instruction could be used when it reduces flag spilling. I had another, more complex case I was going to address in a separate PR.

The other is if similar concerns apply to the SHLX/SHRX transformations. I haven't checked the code size on those but part of the reason I did this in the way I did was looking at those.

I am also hoping to do some kind of more general testing to see what the impact is outside of the LLVM tests. I'll definitely check on my codebase at work which is performance-sensitive. I also recall seeing someone with an automated performance testing system somewhere.

Thanks!

**Comment by KanRobert** [comment]
> I don't know that there is a benefit to eliminating a MOV in most cases. Obviously depending on REX and whatever but I think it's generally worse?

The total cycle of MOV+SHR is same as RORX https://godbolt.org/z/z98vevYMq
The `MOV+SHR` has longer size only when both of two registers are R8-R15. But `MOV` itself does have cost, it's hard to say which is better w/o testing.

> Changing and 8bit or 16bit shift to a 32bit RORX could have a false dependency so I'd definitely remove those. Thanks for pointing that out. Otherwise it should be fine?

32bit/64bit can introduce false dependency too.  Considering the value of source register is `00 00 00 ff 44 33 22 11`, if we shift right by 8, then the high 32 bits are zeros, but if we use rotate, then value would be `11 00 00 00 ff 44 33 22`, no user of the high 32 bits but it's not zero.

> I have two questions, first is where this should better be done. I also think there may be other situations where a normally less optimal instruction could be used when it reduces flag spilling.

I haven't had a clear answer for this. It might be in `X86DAGToDAGISel::tryShiftAmountMod` or in peephole optimization.

File: llvm/lib/Target/X86/X86ISelDAGToDAG.cpp [hunk 3]
```diff
@@ -4212,6 +4213,89 @@ MachineSDNode *X86DAGToDAGISel::emitPCMPESTR(unsigned ROpc, unsigned MOpc,
   return CNode;
 }
 
+// When the consumer of a right shift (arithmetic or logical) wouldn't
+// notice the difference if the instruction was a rotate right instead
+// (because the bits shifted in are truncated away), the shift can be
+// replaced by the RORX instruction from BMI2. This doesn't set flags and
+// can output to a different register. This increases code size in most
+// cases, can have a false dependency when promoting the operand size,
```
Comment (topperc) [inline llvm/lib/Target/X86/X86ISelDAGToDAG.cpp:4221]: Every Intel CPU that implements RORX renames AL, AX, EAX, and RAX together. Only the H registers can be separated. Converting an 8-bit read to a 16-bit read may create a false dependency on one of the H registers, but I'm not sure any of the other cases do.

File: llvm/lib/Target/X86/X86ISelDAGToDAG.cpp [hunk 4]
```diff
@@ -4212,6 +4213,96 @@ MachineSDNode *X86DAGToDAGISel::emitPCMPESTR(unsigned ROpc, unsigned MOpc,
...
+  // similar heuristic as used in the conversion to LEA, namely looking at the
+  // operand for an instruction that creates flags where those flags are used.
+  // This will have both false positives and false negatives. Ideally, both of
+  // these happen later on. Perhaps in copy to flags lowering or in register
+  // allocation.
+  bool MightClobberFlags = false;
+  SDNode *Input = N->getOperand(0).getNode();
+  for (auto Use : Input->uses()) {
+    if (Use->getOpcode() == ISD::CopyToReg) {
+      auto RegisterNode =
```
Comment (topperc) [inline llvm/lib/Target/X86/X86ISelDAGToDAG.cpp:4282]: `auto *RegisterSDNode`  to make it clear it's a pointer.

File: llvm/lib/Target/X86/X86ISelDAGToDAG.cpp [hunk 5]
```diff
@@ -4216,6 +4217,97 @@ MachineSDNode *X86DAGToDAGISel::emitPCMPESTR(unsigned ROpc, unsigned MOpc,
...
+    OpSize = 16;
+  else if (VT == MVT::i8)
+    return false; // i8 shift can't be truncated.
+  else
+    llvm_unreachable("Unexpected shift size");
+
+  unsigned TruncateSize = 0;
+  // This only works when the result is truncated.
+  for (const SDNode *User : N->uses()) {
+    auto name = User->getOperationName(CurDAG);
```
Comment (RKSimon) [inline llvm/lib/Target/X86/X86ISelDAGToDAG.cpp:4254]: unused variable

File: llvm/lib/Target/X86/X86ISelDAGToDAG.cpp [hunk 6]
```diff
@@ -4216,6 +4217,97 @@ MachineSDNode *X86DAGToDAGISel::emitPCMPESTR(unsigned ROpc, unsigned MOpc,
...
+      auto *RegisterNode =
+          dyn_cast<RegisterSDNode>(Use->getOperand(1).getNode());
+      if (RegisterNode && RegisterNode->getReg() == X86::EFLAGS) {
+        MightClobberFlags = true;
+        break;
+      }
+    }
+  }
+  if (!MightClobberFlags)
+    return false;
```
Comment (RKSimon) [inline llvm/lib/Target/X86/X86ISelDAGToDAG.cpp:4296]: Is this correct? The logic appears to be flipped.

File: llvm/lib/Target/X86/X86ISelDAGToDAG.cpp [hunk 7]
```diff
@@ -4216,6 +4217,97 @@ MachineSDNode *X86DAGToDAGISel::emitPCMPESTR(unsigned ROpc, unsigned MOpc,
...
 
+// When the consumer of a right shift (arithmetic or logical) wouldn't notice
+// the difference if the instruction was a rotate right instead (because the
+// bits shifted in are truncated away), the shift can be replaced by the RORX
+// instruction from BMI2. This doesn't set flags and can output to a different
+// register. However, this increases code size in most cases, and doesn't leave
+// the high bits in a useful state. There may be other situations where this
+// transformation is profitable given those conditions, but currently the
+// transformation is only made when it likely avoids spilling flags.
+bool X86DAGToDAGISel::rightShiftUncloberFlags(SDNode *N) {
```
Comment (RKSimon) [inline llvm/lib/Target/X86/X86ISelDAGToDAG.cpp:4228]: typo: rightShiftUncloberFlags -> rightShiftUnclobberFlags

**Review by RKSimon — CHANGES_REQUESTED** [review]

File: llvm/lib/Target/X86/X86ISelDAGToDAG.cpp (hunk 6 above)
Comment (Bryce-MW) [reply llvm/lib/Target/X86/X86ISelDAGToDAG.cpp:4294]: It should be correct? I've clarified the names / explanation a bit but it's possible that I got the logic wrong

**Comment by Bryce-MW** [comment]
I think the fail on Windows is not related. Hopefully a merge fixes it...

File: llvm/lib/Target/X86/X86ISelDAGToDAG.cpp [hunk 8]
```diff
@@ -4216,6 +4217,95 @@ MachineSDNode *X86DAGToDAGISel::emitPCMPESTR(unsigned ROpc, unsigned MOpc,
...
+  // If the shift argument has non-dead EFLAGS, then this shift probably
+  // clobbers those flags making the transformation to RORX useful. This may
+  // have false negatives or positives so ideally this transformation is made
+  // later on.
+  bool ArgProducesFlags = false;
+  SDNode *Input = N->getOperand(0).getNode();
+  for (auto Use : Input->uses()) {
+    if (Use->getOpcode() == ISD::CopyToReg) {
+      auto *RegisterNode =
+          dyn_cast<RegisterSDNode>(Use->getOperand(1).getNode());
```
Comment (RKSimon) [inline llvm/lib/Target/X86/X86ISelDAGToDAG.cpp:4285]: ```dyn_cast<RegisterSDNode>(Use->getOperand(1))```

**Comment by Bryce-MW** [comment]
I spent some time trying out something much more complex: starting at the user of flags that has other inputs (ADC, SBB, CMOVcc are the main ones), trace back the non-flags inputs to see if the node producing the flags inputs is along their paths then check the path from there to the flags user for instructions that produce flags and check if they can be rewritten. This works, but I felt like it was too complicated, , isn't particularly efficient, and didn't seem to improve any code that I tested with.

I have some ideas for future PRs related to avoiding flags spilling so if I come up with a better way to do this kind of thing in the future, I can always come back to it.

**Comment by Bryce-MW** [comment]
As it is, this optimization is very rare. Outside of the function I was optimizing, I didn't see any other instances on my codebase at work. It looks like there [aren't any changes on other tests either](https://llvm-compile-time-tracker.com/compare.php?from=275729ae06d568e9589392c142a416fb8c2bb1a8&to=8cb180bc7b1f5e19358635d6aad756c0c090fb9e&stat=size-text). The only case I can really think of where this transformation happens is 1s complement folding (i.e. internet checksum calculation which is my use) and all the implementations of that that I have seen use inline assembly or a different (slightly) less efficient implementation.

I feel like it is still worthwhile to include since there is no other way (other than inline assembly) to convince the compiler to make this transformation.

**Comment by Bryce-MW** [comment]
I think that I am going to close this for now. Doing this during instruction selection is nice because of load folding at least but it makes it harder to find situations where the optimization can be done, especially if I started adding other similar transformations. I think what I'd like to do is instead work on doing things like this in the copy flags lowering stage.

**Pull request closed without merging**